src/connection.js

```diff
@@ -717,10 +731,33 @@ class Connection extends EventEmitter {
     });
   }
 
+  // Returns false to apply backpressure.
   sendDataToTokenStreamParser(data) {
     return this.tokenStreamParser.addBuffer(data);
   }
 
+  // This is an internal method that is called from Request.pause().
+  // It has to check whether the passed Request object represents the currently
+  // active request, because the application might have called Request.pause()
+  // on an old inactive Request object.
+  pauseRequest(request) {
+    if (this.isRequestActive(request)) {
+      this.tokenStreamParser.pause();
```

Comment:
OK, I see your point. It would be the same kind of check as in Connection.cancel(). But it's problematic.

Consider the following sequence:

- The application calls connection.execSql() to start a query.
-  The query takes longer than expected. A request timeout occurs. The driver sends ATTENTION to the server and enters the SENT_ATTENTION state. The application has no way of knowing this.
- While the driver is in the SENT_ATTENTION state and waiting for the attention response from the server, the application calls Connection.cancel(). The current implementation of the cancel() method considers this an error, writes the debug message "_Requests can only be canceled in the SentClientRequest state, not the SentAttention state_" and returns false.

From an API point of view, I would rather be forgiving than fussy. When the application wants to pause or resume a request that has already completed, then it's OK, because there are no more rows anyway that would be affected. Also when the application wants to cancel a request that has already been canceled, or close a file that is already closed, it's common to just ignore it. The driver has to protect it's inner state and the consistency of the API and therefore ignore pause/resume calls when they are not called with the currently active Request object. But it's not the purpose of the driver to educate the application.